infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/PreviewUrlEndpointsProvisioner.java

```diff
@@ -0,0 +1,127 @@
+import static org.eclipse.che.workspace.infrastructure.kubernetes.server.KubernetesServerExposer.SERVER_UNIQUE_PART_SIZE;
+
+import io.fabric8.kubernetes.api.model.IntOrString;
+import io.fabric8.kubernetes.api.model.Service;
+import io.fabric8.kubernetes.api.model.ServicePort;
+import io.fabric8.kubernetes.api.model.extensions.HTTPIngressPath;
+import io.fabric8.kubernetes.api.model.extensions.Ingress;
+import io.fabric8.kubernetes.api.model.extensions.IngressBackend;
+import io.fabric8.kubernetes.api.model.extensions.IngressRule;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.eclipse.che.api.core.model.workspace.runtime.RuntimeIdentity;
+import org.eclipse.che.api.workspace.server.model.impl.CommandImpl;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesEnvironment;
+import org.eclipse.che.workspace.infrastructure.kubernetes.server.ServerServiceBuilder;
+import org.eclipse.che.workspace.infrastructure.kubernetes.server.external.ExternalServerExposer;
+import org.eclipse.che.workspace.infrastructure.kubernetes.util.Services;
+
+/**
+ * For Commands that have defined Preview URL, tries to find existing {@link Service} and {@link
+ * Ingress}. When not found, we create new ones and put them into given {@link
+ * KubernetesEnvironment}.
+ *
+ * @param <T> type of the environment
+ */
+@Singleton
+public class PreviewUrlEndpointsProvisioner<T extends KubernetesEnvironment>
+    implements ConfigurationProvisioner<T> {
+
+  private final ExternalServerExposer<T> externalServerExposer;
+
+  @Inject
+  public PreviewUrlEndpointsProvisioner(ExternalServerExposer<T> externalServerExposer) {
+    this.externalServerExposer = externalServerExposer;
+  }
+
+  @Override
+  public void provision(T k8sEnv, RuntimeIdentity identity) throws InfrastructureException {
+    createEndpointsForPreviewUrls(k8sEnv);
+  }
+
+  private void createEndpointsForPreviewUrls(T env) {
+    List<CommandImpl> previewUrlCommands =
+        env.getCommands()
+            .stream()
+            .filter(c -> c.getPreviewUrl() != null)
+            .collect(Collectors.toList());
+
+    List<ServicePort> portsToProvision = new ArrayList<>();
+    for (CommandImpl command : previewUrlCommands) {
+      int port = command.getPreviewUrl().getPort();
+      Optional<Service> foundService =
+          Services.findServiceWithPort(env.getServices().values(), port);
+      if (foundService.isPresent()) {
+        if (!hasMatchingEndpoint(env, foundService.get(), port)) {
+          externalServerExposer.expose(
+              env,
+              null,
+              foundService.get().getMetadata().getName(),
+              createServicePort(port),
```

Comment:
Can you point where you provision this created port into service?

Reply:
now using the found port from the service which is already provisioned